### ISSUE TITLE
We want a "Partners" page that looks like Luminate's

### ISSUE THREAD

**Issue by danblah** [ResetNetwork/apply-app#79]
see https://luminategroup.com/portfolio.

- [ ] Create a new page type for Partners (start with the fields in the above example)
- [ ] Create a new index page for Partners that lists them all (style in the example above)
- [ ] Add filtering capability
- [ ] Add search capability

**Comment by scott-psycle** [comment]
@danblah We'll start to have a look at spec'ing the details for this new page type.  A couple of initial questions though.

1. Am I correct that you want to have a list page for ALL investments and also an individual Partner page type which gives details of a specific partner organisation, similar to this on Luminate?  https://luminategroup.com/investee/africa-check

2. On Luminate we have a couple of entities:

* Investment
* Organisation (Partner)

An Organisation can have multiple investments associated with it.  Will this be the case with Reset?  Or will it be 1:1?

I'll come back to you to confirm specifics about data fields and values once we've had a chance to discuss.

**Comment by danblah** [comment]
1. Yes.
2. Yes. 

Exactly.

**Comment by danblah** [comment]
Also not for MVP but so we don't make it in a way that makes it not possible... Along with the ability to manually create organizations and add projects/investments to them, we're going to want the ability to create a new project/investment from approved applications (or associate a new application to an existing organization as a new investment) in an automated way.

**Comment by danblah** [comment]
Bump bump on this. Can we get an update and some progress here?

**Comment by Parbhat** [comment]
@danblah Is this issue ready to work on? If yes and no one is working on it, @ankitjavalkar can start work on it.

**Comment by frjo** [comment]
@Parbhat Yes, this is ready to work on.

**Comment by Parbhat** [comment]
Thanks @frjo. @ankitjavalkar You can go through the issue and ask questions if need more details in requirements.

**Comment by frjo** [comment]
@Parbhat @sks444 Some parts of this might be time-consuming. Reset would not mind a simpler version with the basic functionality as a start. Extra features can then be added in future work.

The two content types with a minimum set of fields and a table view is a good start. Maybe possible to reuse part from "/apply/submissions/all/" able view?

**Comment by sks444** [comment]
Thanks for the updates @frjo. I will look into the table implementations done for `/apply/submissions/all/` and try to use the same for this feature.

**Comment by sks444** [comment]
@frjo, wanted to confirm a few things before proceeding with the implementations:

- [ ] We need to create two entities `Organization`(partner) and `Investment`. (One organization can have many investments)

- [ ] For the initial version as in MVP, these entities will be manually created from the admin portal.

- By looking at the Luminate's data I have come up with following fields for the initial version, let me know if there is more to be added:

```
Organization:
    - name 
    - status: current partner, alumni partner
    - type: profit, non-profit
    - description: a small description about the partner
    - logo-url: partner logo
    - web-url: website url of the partner

Investment:
    - name
    - primary impact area: a choice field with choices(Civic Empowerment, Data & Digital Rights, Financial Transparency, Independent Media, Reset, Other)
    - geographic focus: a choice field with choices(Africa, Central & Eastern Europe, Global, Latin America, Multi-Country, Southeast Asia, United States, Western Europe, other)
    - year
    - amount committed(in USD)
    - type
    - purpose
    - description
```

- [ ] Should the fields `primary impact area` and `geographic focus` be at Organization level or having them at Investment level is okay? 

- [ ] Should the fields `primary impact area` and `geographic focus` be a text input field or a choice field with above given choices is okay?

- [ ]  Should the field investment `type` be a choice field or a text input field. If choice, what could be the choices? Luminate's have values like, `Grant: Foreign Public Charity Equivalent`, `Capital Budget` etc.

- [ ] Should the field investment `purpose` be a choice field or a text input field. If choice, what could be the choices? Luminate's have values like, `General Funding Grant`, `Project Grant`, `Program Grant`, etc.

**Comment by frjo** [comment]
@danblah Can you answer the questions above?

**Comment by sks444** [comment]
@frjo, setting up the this repo for the first time, while I know that the setup and everything is very similar to hypha. 

When I run the server locally it still gives me sites configured for Open Technology Fund(E.g. Homepage etc.)

Is there some configuration that we need to do to have the sites configured with reset interface locally?

Also, please let me know if there is a db dump that you have specific to this repo.

Or if there is a document on doing specific setup for reset.

**Comment by frjo** [comment]
@sks444  Reset has custom models for front page etc. Will get you a sanitised db dump from Reset test, that will be the quickest and easiest I think.

**Comment by sks444** [comment]
Yes, that will do the work. Thank you.

**Comment by danblah** [comment]
@sks444 and @frjo, responding.

- [ ] We need to create two entities `Organization`(partner) and `Investment`. (One organization can have many investments)

**Yes, let us call it partner.**

- [ ] For the initial version as in MVP, these entities will be manually created from the admin portal.

**Yes, that's right.**

- By looking at the Luminate's data I have come up with following fields for the initial version, let me know if there is more to be added:

**I've modified below.**


```
Partner ~~Organization~~:
    - name*
    - status: active, inactive* ~~current partner, alumni partner~~
    - public: yes/no* 
    - description: a small description about the partner
    - logo-url: partner logo
    - web-url: website url of the partner
    - Option to include and require one or more Category Question, for example:
    -- type: Can we use the Category Question "[Organization Type](https://apply.reset.tech/admin/categories/category/edit/1/)"? ~~profit, non-profit, individual~~

Investment:
    - name*
    - year*
    - amount committed (in USD)*
    - description*
    - application: select from any approved application
    - Option to include and require one or more Category Question, for example:
    -- type of work: Can we use the Category Question "[Type of Work](https://apply.reset.tech/admin/categories/category/edit/5/)"? ~~primary impact area: a choice field with choices(Civic Empowerment, Data & Digital Rights, Financial Transparency, Independent Media, Reset, Other)~~
    -- region: Can we use Category Question "[Region]([url](https://apply.reset.tech/admin/categories/category/edit/4/))"? ~~geographic focus: a choice field with choices(Africa, Central & Eastern Europe, Global, Latin America, Multi-Country, Southeast Asia, United States, Western Europe, other)~~

* indicates required

```

- [ ] Should the fields `primary impact area` and `geographic focus` be at Organization level or having them at Investment level is okay? 

**Investment level is right and okay. This assumes we would easily be able to search/filter/find organisations by fields of the investments associated with them.** 

- [ ] Should the fields `primary impact area` and `geographic focus` be a text input field or a choice field with above given choices is okay?

**They should be replaced by choosing a Category Question.**

- [ ]  Should the field investment `type` be a choice field or a text input field. If choice, what could be the choices? Luminate's have values like, `Grant: Foreign Public Charity Equivalent`, `Capital Budget` etc.

**They should be replaced by choosing a Category Question.**


- [ ] Should the field investment `purpose` be a choice field or a text input field. If choice, what could be the choices? Luminate's have values like, `General Funding Grant`, `Project Grant`, `Program Grant`, etc.

**They should be replaced by choosing a Category Question.**

**Comment by frjo** [comment]
@sks444  We are moving to have this feature built in to Hypha directly so moving issue too the Hypha repo.

**Comment by frjo** [comment]
Hmm, thought you could move issues to any repo you had access to but it looks like it limited to repos within the same organisation? Anyone knows how to move it to OTF/hypha repo?

**Comment by Parbhat** [comment]
@frjo Looks like we can not transfer issues to different organisation. Maybe we can keep the issue here and @sks444 can create PR in Hypha repo.

**Comment by Parbhat** [comment]
@frjo Creating this feature on Hypha repo means we will need style changes both for OTF and Reset. As these pages will be on public site and both OTF and Reset have different design.

**Comment by frjo** [comment]
@Parbhat Yes, Reset will as a minimum need their own template for this page.

**Comment by frjo** [comment]
@danblah This can be tested at https://test.opentech.fund/ 

See https://github.com/OpenTechFund/hypha/pull/2126 for how and what to test.

**Comment by danblah** [comment]
**Modifications to what we have right now...**
_Index page_
- Use Partner instead of Partner name in the listing header
- When creating a new Partner, the Name/Title fields are redundant. Can we remove one (ie rename Title to Name)?
- On the index page (ie https://test.opentech.fund/investments/), can we add a comma and remove the decimal (ie 100,00 not 100000.00)?
- On the index page, use these fields only: Partner name, year, status, amount.
- Can we change the index page Introduction field to a Richtext editor used in the Partner Description field?
- Can we control the slug (ie /about/portfolio)? Seems its hardcoded to /investments?

_Partner page_
- Can we remove the Public: yes/no from showing up on the partner page?
- Does the public yes/no affect visibility of a partner page (and associated investments)?

**Things to add...**
_Investment page_
- Can add a Category field, just like we do with the application edit form (or something similar), where we can choose to add any one of the Category Questions (https://test-apply.opentech.fund/admin/categories/category/) to the form, and then select one when creating a new record? See https://luminategroup.com/investee/all-out, where Impact Area, Geographic Focus, etc would all be Categories with choices.

_Partners page_
- When Category Questions above are being used, can we automatically list them on the partner page based on the selected options in associated investments? See https://luminategroup.com/investee/all-out where for example, if an investment associated with partner All Out selected Latin America for the category question Geographic focus, that would show on the Partner page.
- Can we create a "Funding details" section at the bottom, listing investments like this https://luminategroup.com/investee/all-out, where the name in that is the name/title of the investment, each listing expands to show the investment description, and there is a sum of all the amounts?

_Index page_
- On the index page, can we expand the listing (like we do with submissions) to display the Investment Description (see how green x at the far right of a listing works at https://luminategroup.com/portfolio)?
- Can the default sort by year (and most recently created investment)?
- Search by partner name box
- Can we automatically add a header for any Category Questions used in investments?
- Ability to filter by year, status, and amount
- Ability to sort the list by fields (name, year, amount, etc)
- Add a "Records per page" ie 25, 50, 100 with pagination (ie page 1, 2, 3, next etc)
- Add a "Download data (csv)" that pulls presents the user a csv of all investments to download.

**Even more general...**
- Can we get this on test.reset.tech next so we can start design debugging?
- Do we need to change how we're working with years or other fields to filter results (go to https://luminategroup.com/portfolio, click filter, see filters)? For example, do we need to pre-define years somewhere first?

**Comment by sks444** [comment]
@danblah, thanks for the feedback. Will improve the current functionalities and the new ones.

**Regarding category questions:** Can we have a fixed number fields with fixed number of choices(_see image)?

This feature works very differently than application forms. We should consider these points before building the `category questions` feature for this:

-  Building this would be very time consuming in comparison to having fixed number of fields and choices.
-  Are we gonna add new category questions or change choices very frequently? If not, we can always create a small pr to add it.
- Would not be easy building a table with dynamic header and the ui would also not be consistent.
- Showing filters for dynamic fields/choices and search can also lead to complications.

I think Luminate might be doing something similar and have the following number of fields with constant choices:

<img width="273" alt="Screenshot 2020-09-16 at 11 40 19 AM" src="https://user-images.githubusercontent.com/22263040/93300020-aa182e80-f813-11ea-9046-67ebadeec30a.png">

Dan, can you please confirm once:

- [ ] if we should go ahead with building `category questions` for this feature

 **Or**

- [ ] It is okay to have the above fields with their choices(can modify those based on your suggestions)? At least for the MVP version of it.

**Comment by scott-psycle** [comment]
@danblah Do you need our assistance implementing the Reset front end?

**Comment by danblah** [comment]
heya @sks444! 

* Category Questions will change regularly, be completely different from implementation to implementation (ie Reset's are different than OTF's, etc), and we want these to continue being fully manageable by users.
* Category Question options/choices will also change (although are not deleted/removed, they are "archived").
* I think one reason to use Category Questions is based on the desire to make it very easy and/or automate the creation of a Partner and Investment when an application is approved, thus ensuring all things are labelled from the same source of labels.
* Another reason is the desire to expand the API enabling a query to category questions and a response with all matching content available (ie query for geographic focus "United States" and the response is all investments, partners, projects, applications, with that label), again wanting to ensure all things are labelled from the same source of labels.

How do we reconcile the need for a barebones mvp short-term and these needs mid-term? Is there a small function you can build to generate the fixed fields and choices from existing Category Questions now? Is there a hook for changes to Category Questions you can use to regenerate these fields whenever Category Questions change? Good/bad ideas?

Can you lay out a new issue for meeting the mid-term needs above as a more permanent solution?

Thanks again!

**Comment by danblah** [comment]
@scott-psycle, as @sks444 gets all the values spitting out, it'd be great for you to take on refining the look/feel and extending any of the design templates as needed. @sks444 and @frjo, y'all sync up with @scott-psycle to figure out where the right hand-off is?

**Comment by sks444** [comment]
Thanks for the inputs @danblah, that is very helpful!

I had a discussion with @Parbhat. We have a plan and we are going to try out the implementation for dynamically creating category questions and their choice selection for this feature.

**Comment by sks444** [comment]
Hi @scott-psycle,

Yes, once we are done with implementing this in hypha. We shall discuss on having this feature in reset and making the design better. We will let you know once this is done. Thanks.
cc: @frjo

**Comment by sks444** [comment]
Hi all, a good amount of work has been done on this. Here are the updates to get further feedback:

I will quote the @danblah [comments](https://github.com/ResetNetwork/apply-app/issues/79#issuecomment-692903040) on requested changes and make a tick mark if that is done. And will also add a small description on the solution provided if needed.

**Modifications to what we have right now...**
_Index page_

- [ ] Use Partner instead of Partner name in the listing header 
- [ ] When creating a new Partner, the Name/Title fields are redundant. Can we remove one (ie rename Title to Name)?
> Now we have `Title` instead of having both `Name` and `Title`. `Title` field is coming from wagtail, so overriding it to `Name` would be a lot of unnecessary work.
- [ ] On the index page (ie https://test.opentech.fund/investments/), can we add a comma and remove the decimal (ie 100,00 not 100000.00)?
- [ ] On the index page, use these fields only: Partner name, year, status, amount.
- [ ] Can we change the index page Introduction field to a Richtext editor used in the Partner Description field?
- [ ] Can we control the slug (ie /about/portfolio)? Seems its hardcoded to /investments?

_Partner page_
- [ ]  Can we remove the Public: yes/no from showing up on the partner page?
- [ ]  Does the public yes/no affect visibility of a partner page (and associated investments)?
> Yes. Investments for partners which are not public won't be listed on the table and also the partner page for a private partner will be 404.

**Things to add...**
_Investment page_

- [ ] Can add a Category field, just like we do with the application edit form (or something similar), where we can choose to add any one of the Category Questions (https://test-apply.opentech.fund/admin/categories/category/) to the form, and then select one when creating a new record? See https://luminategroup.com/investee/all-out, where Impact Area, Geographic Focus, etc would all be Categories with choices.
> Yes. This is possible now. See the section below on, **Adding category questions to the investments.**

_Partners page_

- [ ] When Category Questions above are being used, can we automatically list them on the partner page based on the selected options in associated investments? See https://luminategroup.com/investee/all-out where for example, if an investment associated with partner All Out selected Latin America for the category question Geographic focus, that would show on the Partner page.
- [ ]  Can we create a "Funding details" section at the bottom, listing investments like this https://luminategroup.com/investee/all-out, where the name in that is the name/title of the investment, each listing expands to show the investment description, and there is a sum of all the amounts?

_Index page_

- [ ] On the index page, can we expand the listing (like we do with submissions) to display the Investment Description (see how green x at the far right of a listing works at https://luminategroup.com/portfolio)?
- [ ] Can the default sort by year (and most recently created investment)?
> Default sort is by most recently created investment.

- [ ]  Search by partner name box
- [ ]  Can we automatically add a header for any Category Questions used in investments?
- [ ] Ability to filter by year, status, and amount
- [ ] Ability to sort the list by fields (name, year, amount, etc)
- [ ] Add a "Records per page" ie 25, 50, 100 with pagination (ie page 1, 2, 3, next etc)
- [ ] Add a "Download data (csv)" that pulls presents the user a csv of all investments to download.

**Even more general...**
- [ ]  Can we get this on test.reset.tech next so we can start design debugging?
> Tagging @frjo for this.

- [ ] Do we need to change how we're working with years or other fields to filter results (go to https://luminategroup.com/portfolio, click filter, see filters)? For example, do we need to pre-define years somewhere first?
> No. We will only show the years in the filter for which the investments is made. And not redundant years for which there is no investments.

### Adding Category Questions To Investments

- Visit `Apply-> Category Questions` in admin. And create category questions that you need for investments.
- `Settings-> Investment Category Settings`:  Select the category questions that you want to use for investments. Make sure to add it for both the sites(I.e. `apply` and `public`)
- Investments-> Add Investment: Now the form will also have the category questions that you added. Select the valid choices and create investment.
- Visit the investment table and there you would see the category questions is automatically added to the columns.

**Comment by sks444** [comment]
@danblah this is deployed to test with the new updates.

**Comment by danblah** [comment]
@sks444, just testing things out... super impressed.

we're nearly to mvp!

Im getting some 500 errors when i try to edit existing records btw, see https://test-apply.opentech.fund/admin/partner/investment/edit/4/. I think there's some bug where multiple partners are being created as well, see https://test.opentech.fund/results/new-partner-page/partner/ and https://test.opentech.fund/results/new-partner-page/partner/, which i think should be the same partner?

On the portfolio page (see https://test.opentech.fund/about/portfolio/)
* Replace "Investment Description" with the Name field for the investment
* Include the category questions, if they are set when you expand the investment detail

On partner page (see https://test.opentech.fund/results/new-partner-page/test-partner-a/)
* Can we show all the terms from any category used for all associated investments? Ie partner has one investment focused on country brazil and another for country chile. On the partner page, we'd see under Country: Brazil, Chile
* Under total investment, can we list all the investments associated with that partner, sorted by year and expandable just like the portfolio page? See https://luminategroup.com/investee/digital-freedom-fund
* 

That will get us to a production MVP for styling on reset.tech I think.

I also need a page that outputs Partner logo and name in a responsive grid, default to only showing those that are active, with the ability to filter for all or only inactive.

I also need the ability to add a Partners block on our front page that defaults to showing the most recently invested in partners (but also allows us to promote specific ones to the front).

**Comment by sks444** [comment]
Hi @danblah,

Can you do a quick testing on this again?

- [ ] Note that listing investments under a partner page is not done yet.

> Include the category questions, if they are set when you expand the investment detail

As per the old requirements we automatically create a column in the investments table for category questions that gets added for investments. 

> ### Adding Category Questions To Investments
> - Visit `Apply-> Category Questions` in admin. And create category questions that you need for investments.
> - `Settings-> Investment Category Settings`:  Select the category questions that you want to use for investments. Make sure to add it for both the sites(I.e. `apply` and `public`, _There is an option to select site while editing/creating settings in the top-right_.)
> - Investments-> Add Investment: Now the form will also have the category questions that you added. Select the valid choices and create investment.
> - Visit the investment table and there you would see the category questions is automatically added to the columns. 

I have already added some test data for this while testing..we can ignore the `Test Partner 1` listed at https://test-apply.opentech.fund/admin/pages/926/, I believe that is a cached page which doesn't exist in the db..we can fix that whenever we redeploy https://github.com/HyphaApp/hypha/pull/2126.


Let us try to get this merged in this iteration..as it's getting harder to keep track of things with this big of a stale pr.

**Comment by danblah** [comment]
Merge it!

**Comment by sks444** [comment]
Thanks @danblah.

@frjo can we redeploy https://github.com/HyphaApp/hypha/pull/2126 to test..but first we would have to make sure that no data from the old migration exist..I guess we may need to follow steps from https://github.com/HyphaApp/hypha/pull/2126#issuecomment-803902944 in order to delete `Test partner 1`.

Once that is deployed..I can do a quick testing to make sure that the migrations are not causing any issues..then we can go-ahead and merge this.

**Comment by frjo** [comment]
@sks444 Now redeployed to https://test-apply.hypha.app/ (OBS! the Hypha test site, cloned the DB from OTF test so same accounts etc.)

**Comment by sks444** [comment]
Thanks @frjo.

It seems to be working fine.

We can merge this in and create new issue for the features that remain:

from https://github.com/ResetNetwork/apply-app/issues/79#issuecomment-790857792 : 
> - [ ] List investments table under a partner page
> - [ ] I also need a page that outputs Partner logo and name in a responsive grid, default to only showing those that are active, with the ability to filter for all or only inactive.

> - [ ]  I also need the ability to add a Partners block on our front page that defaults to showing the most recently invested in partners (but also allows us to promote specific ones to the front).

**Comment by frjo** [comment]
The PR https://github.com/HyphaApp/hypha/pull/2126 is now merged.

Added three new issues, first to Hypha and the last two to Reset as they seem Reset specific.
